symengine/dense_matrix.cpp

```diff
@@ -364,76 +364,96 @@ void mul_dense_scalar(const DenseMatrix &A, const RCP<const Basic> &k,
 }
 
 // ---------------------------- Joining Operations ---------------------------//
-void row_join(const DenseMatrix &A, const DenseMatrix &B, DenseMatrix &C)
+void DenseMatrix::row_join(const DenseMatrix &B)
 {
-    SYMENGINE_ASSERT(A.row_ == B.row_
-                     and (A.row_ == C.row_ and C.col_ == A.col_ + B.col_));
+    this->col_insert(B, col_);
+}
 
-    unsigned row = A.row_, col = A.col_;
+void DenseMatrix::col_join(const DenseMatrix &B)
+{
+    this->row_insert(B, row_);
+}
 
-    for (unsigned i = 0; i < row; i++) {
+void DenseMatrix::row_insert(const DenseMatrix &B, unsigned pos)
+{
+    SYMENGINE_ASSERT(col_ == B.col_ and pos <= row_)
+
+    unsigned row = row_, col = col_;
+    DenseMatrix C = *this;
```

Comment:
This can be done without a temporary. First, resize `this`, then move the elements of `this` below row `pos` to the end and then insert `B`.